### ISSUE TITLE
Performance regression on CPU from 0.4.1 to 1.0.0 on ResNet inference

### ISSUE THREAD

**Issue by lopuhin** [pytorch/pytorch#16717]
## 🐛 Bug

ResNet50 inference is 20 - 65% slower on pytorch 1.0.0 vs 0.4.1 on CPU with `OMP_NUM_THREADS=1`, and  2800% slower without `OMP_NUM_THREADS=1` (8 core CPU), although these are probably two separate issues and I'm personally more interested in `OMP_NUM_THREADS=1` behaviour because it gets maximal use of available cores regardless of the version used.

## To Reproduce

- install pytorch from PyPI on Linux with Python 3.6, specifically these wheels were installed: `https://files.pythonhosted.org/packages/7e/60/66415660aa46b23b5e1b72bc762e816736ce8d7260213e22365af51e8f9c/torch-1.0.0-cp36-cp36m-manylinux1_x86_64.whl` and `https://files.pythonhosted.org/packages/49/0e/e382bcf1a6ae8225f50b99cc26effa2d4cc6d66975ccf3fa9590efcbedce/torch-0.4.1-cp36-cp36m-manylinux1_x86_64.whl`
- run `CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 ipython`
- in ipython, run:
```
import torch, torchvision
model = torchvision.models.resnet50(num_classes=10)
with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))`
```
- still in ipython, run:
`%timeit with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))`

## Expected behavior

I expected pytorch 1.0.0 to be in the same ballpark as pytorch 0.4.1, or faster. Instead it is from 20% to 65% slower. I tested on two different servers, one old `Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz` and another newer `Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz`.

Results of `%timeit` on `Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz`:

pytorch 0.4.1:
```
In [4]: %timeit with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))
362 ms ± 40.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

pytorch 1.0.0:
```
In [4]: %timeit with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))
596 ms ± 57.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Results on `Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz`:

pytorch 0.4.1:
```
In [6]: %timeit with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))
174 ms ± 786 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

pytorch 1.0.0:
```
In [5]: %timeit with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))
210 ms ± 593 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Also on the same server (8 virtual cores), without ``OMP_NUM_THREADS=1`` I get ``143 ms ± 4.14 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`` for 0.4.1, and ``4.05 s ± 126 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`` for 1.0.0 (so 28 times slower).

## Environment

For pytorch 0.4.1:
```
# python3 collect_env.py
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 9.2.148
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

Same for 1.0.0:

```
# python3 collect_env.py
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 9.2.148
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

**Comment by ailzhang** [comment]
cc: @cpuhrsch

**Comment by sidazhang** [comment]
I am also observing this. The scale looks about right. Suddenly the CPU performance is about between 20x to 100x slower. 

The last time, I benchmarked CPU was on 1.0 release candidate in November and I manually built with MKL_DNN. And now I use pytorch nightly (I assume MKL DNN is built by default on master now) and CPU performance regressed drastically.

**Comment by ezyang** [comment]
1.0.0 regression is partially explainable by the fact that we accidentally disabled OpenMP, but it doesn't explain the single-threaded slowdown.

**Comment by sidazhang** [comment]
@ezyang Btw, I am on pytorch nightly

**Comment by sidazhang** [comment]
Ok. I realized my CPU performance was due to another reason https://github.com/pytorch/pytorch/pull/5294

My remaining CPU performance problem is this. When running a JIT model in CPU, if you run it in the python runtime, it is twice as fast as if you run from the C++ frontend

**Comment by sidazhang** [comment]
CC: @mingfeima here

**Comment by sidazhang** [comment]
Update: my particular issue was due to denormal floats

1. torch.set_flush_denormal(True)
https://github.com/pytorch/pytorch/pull/5294

And by adding exports
2. `export KMP_AFFINITY=granularity=fine,compact,1,0`, `export KMP_BLOCKTIME=1`

**Comment by cpuhrsch** [comment]
@sidazhang - can we consider the regression as resolved for your purposes?

**Comment by lopuhin** [comment]
FWIW steps outlined by @sidazhang in 
https://github.com/pytorch/pytorch/issues/16717#issuecomment-470802536 do not affect the timings for me posted in the first comment. Also only the issue with OMP_NUM_THREADS=1 is relevant for 1.0.1, performance when using all threads is fixed in 1.0.1, thanks! I updated the top comment to reflect this (but it was mostly about OMP_NUM_THREADS=1 in any case).

**Comment by ezyang** [comment]
I'm able to repro the regression comparing master and 0.4.1 binaries.

Hardware: Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz

Master 56c16fe26f84148d5664ac21dd3943c7d3dded86

```
(/home/ezyang/Dev/pytorch-env) ezyang@autobox:~/Dev/pytorch$ CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 python -
m timeit -s 'import torch, torchvision; model = torchvision.models.resnet50(num_classes=10)' 'with torch.no_g
rad(): model(torch.zeros(1, 3, 224, 224))' 
10 loops, best of 3: 252 msec per loop
```

0.4.1 https://files.pythonhosted.org/packages/49/0e/e382bcf1a6ae8225f50b99cc26effa2d4cc6d66975ccf3fa9590efcbedce/torch-0.4.1-cp36-cp36m-manylinux1_x86_64.whl

```
(/home/ezyang/Dev/pytorch-cuda9-env) ezyang@autobox:~/Dev$ CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 python -m timeit -s 'import torch, torchvision; model = torchvision.models.resnet50(num_classes=10)' 'with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))' 
10 loops, best of 3: 207 msec per loop
```

**Comment by ezyang** [comment]
Warning: the offered code example is doing dropout. The performance regression persists after dropout is disabled, but everything is a bit faster when we don't have to generate random numbers.

```
# master 56c16fe
CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 python -m timeit -s 'import torch, torchvision; model = torchvision.models.resnet50(num_classes=10).train(False)' 'with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))' 
10 loops, best of 3: 208 msec per loop

# 0.4.1 binary
(/home/ezyang/Dev/pytorch-cuda9-env) ezyang@autobox:~/Dev$ CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 python -m timeit -s 'import torch, torchvision; model = torchvision.models.resnet50(num_classes=10).train(False)' 'with torch.no_grad(): model(torch.zeros(1, 3, 224, 224))'
10 loops, best of 3: 179 msec per loop
```

**Comment by mingfeima** [comment]
@ezyang is it possible to add a new **tag** to track CPU issues, maybe `CPU` or more specialized, `MKL`, `MKL-DNN`? this would be really helpful for us to classify `performance` and `functional` issues on the CPU side.

**Comment by gchanan** [comment]
@VitalyFedyunin is going to take a look.

**Comment by ezyang** [comment]
@mingfeima I forgot to respond to you here, but we have `module: cpu`, `module: mkl` and `module: mkldnn`. So these are all hopefully good tags for y'all to monitor :)
